qiskit/providers/backend.py

```diff
@@ -229,3 +239,392 @@ class can handle either situation.
+        """
+
+        self._options = self._default_options()
+        self._provider = provider
+        self._coupling_map = None
+        if fields:
+            for field in fields:
+                if field not in self._options.data:
+                    raise AttributeError("Options field %s is not valid for this backend" % field)
+            self._options.update_config(**fields)
+        self._basis_gates_all = None
+        self.name = name
+        self.description = description
+        self.online_date = online_date
+        self.backend_version = backend_version
+
+    @property
+    def instructions(self) -> List[Tuple[Instruction, Tuple[int]]]:
+        """A list of Instruction tuples on the backend of the form ``(instruction, (qubits)``"""
+        return self.target.instructions
+
+    @property
+    def operations(self) -> List[Instruction]:
+        """A list of :class:`~qiskit.circuit.Instruction` instances that the backend supports."""
+        return list(self.target.operations)
+
+    def _compute_non_global_basis(self):
+        incomplete_basis_gates = []
+        size_dict = defaultdict(int)
+        size_dict[1] = self.target.num_qubits
+        for qarg in self.target.qargs:
+            if len(qarg) == 1:
+                continue
+            size_dict[len(qarg)] += 1
+        for inst, qargs in self.target.items():
+            qarg_sample = next(iter(qargs))
+            if qarg_sample is None:
+                continue
+            if len(qargs) != size_dict[len(qarg_sample)]:
+                incomplete_basis_gates.append(inst)
+        self._basis_gates_all = incomplete_basis_gates
+
+    @property
+    def operation_names(self) -> List[str]:
+        """A list of instruction names that the backend supports."""
+        if self._basis_gates_all is None:
+            self._compute_non_global_basis()
+        if self._basis_gates_all:
+            invalid_str = ",".join(self._basis_gates_all)
+            msg = (
+                f"This backend's operations: {invalid_str} only apply to a subset of "
+                "qubits. Using this property to get 'basis_gates' for the "
+                "transpiler may potentially create invalid output"
+            )
+            logger.warning(msg)
+        return list(self.target.operation_names)
+
+    @property
+    @abstractmethod
+    def target(self):
+        """A :class:`qiskit.transpiler.Target` object for the backend.
+
+        :rtype: Target
+        """
+        pass
+
+    @property
+    def num_qubits(self) -> int:
+        """Return the number of qubits the backend has."""
+        return self.target.num_qubits
+
+    @property
+    def coupling_map(self):
+        """Return the :class:`~qiskit.transpiler.CouplingMap` object"""
+        return self.target.coupling_map()
+
+    @property
+    def instruction_durations(self):
+        """Return the :class:`~qiskit.transpiler.InstructionDurations` object."""
+        return self.target.durations()
```

Comment:
`Target.durations()` does nontrivial work, so having this as a `property` could well mask performance pitfalls.  Can we make this a proper method instead, to make it obvious at the call site that it takes time?

Reply:
I was trying to leave these things as `property` methods because they're attributes of the backend. It makes the access patterns more ergonomic imo and consistent with things like `backend.coupling_map`. I tried to compensate for the performance concern here by caching things in the target if they're potentially expensive to construct, so for >1 call in a loop or something the overhead is just to return from a slotted attribute in `Target`. We only construct a fresh object if the target has been modified since the last call (which for a backend like this shouldn't be the case).

Reply:
Fair enough, that's fine by me.  I've been becoming more and more disillusioned by `@property` getter methods recently, but caching is one situation where I'm usually fine with them.